pkg/infrastructure/azure/azure.go

```diff
@@ -676,7 +677,7 @@ func (p *Provider) PostDestroy(ctx context.Context, in clusterapi.PostDestroyerI
 		networkClientFactory: networkClientFactory,
 	})
 	if err != nil {
-		return fmt.Errorf("failed to delete security rule: %w", err)
+		logrus.Warnf("failed to delete security rule: %v", err)
```

Comment:
If we don't fail here, I think we will end up missing any legitimate errors (i.e. failure to delete in public cluster) and could ship bugs. Also, if we don't skip this delete, it seems like we could potentially delete a NSG we didn't create? The NSG would have to have the same naming conventions, and so is extremely unlikely, but still--probably better to skip.

As you mention, we don't have the installconfig at this point. We do have the cluster metadata, but adding a field to that is my least preferred option (it seems to cause a lot of issues to hive and in general I think it's better to discover resources on known heuristics). 

I think a viable option would be to do something similar to what we are doing in AWS here:
https://github.com/openshift/installer/blob/3ca43c1454d8922dd87951fbcea5563df330621b/pkg/infrastructure/aws/clusterapi/aws.go#L230-L240

- In the destroy step, before we have torn down the CAPI control plane, determine if the cluster is private by reading a value from the CAPI objects (see below). Then save a field on the pointer receiver
- In the post destroy step read that field then skip accordingly

So if we do that, the question is what value to look at? Looking at this PR it seems like the only value we are changing is whether the bootstrap node has a public ip. So that would be fine. The AWS code example looks at the cluster object, so we'd need to change that to read the bootstrap node. If looking at the public IP seems flimsy, or something, we could potentially set a different, more deliberate value, such as changing the name of the outbound LB or something.

Reply:
> If we don't fail here, I think we will end up missing any legitimate errors (i.e. failure to delete in public cluster) and could ship bugs. Also, if we don't skip this delete, it seems like we could potentially delete a NSG we didn't create? The NSG would have to have the same naming conventions, and so is extremely unlikely, but still--probably better to skip.
> 
> As you mention, we don't have the installconfig at this point. We do have the cluster metadata, but adding a field to that is my least preferred option (it seems to cause a lot of issues to hive and in general I think it's better to discover resources on known heuristics).
> 
> I think a viable option would be to do something similar to what we are doing in AWS here:
> 
> https://github.com/openshift/installer/blob/3ca43c1454d8922dd87951fbcea5563df330621b/pkg/infrastructure/aws/clusterapi/aws.go#L230-L240
> 
>     * In the destroy step, before we have torn down the CAPI control plane, determine if the cluster is private by reading a value from the CAPI objects (see below). Then save a field on the pointer receiver
> 
>     * In the post destroy step read that field then skip accordingly
> 
> 
> So if we do that, the question is what value to look at? Looking at this PR it seems like the only value we are changing is whether the bootstrap node has a public ip. So that would be fine. The AWS code example looks at the cluster object, so we'd need to change that to read the bootstrap node. If looking at the public IP seems flimsy, or something, we could potentially set a different, more deliberate value, such as changing the name of the outbound LB or something.

I went the route of specially naming the outbound load balancer when the cluster is private.

Reply:
> I went the route of specially naming the outbound load balancer when the cluster is private.

Cool! I think this is the better route :+1:

Reply:
Apparently we break things doing it this way. The name of the load balancer is expected to be the cluster ID. I will find another work around.

Reply:
I don't see why we just don't check if the rule exists, and if so, delete it and the nat rule. Makes sense to me. I will just include the cluster ID in the rule name.

Reply:
+ 1 for latest approach to make the rule names easily identifiable with clusterID in it. Simple and effective.